client/packages/programs/src/JsonForms/common/components/Array/KeyedItemArray.tsx

```diff
@@ -0,0 +1,134 @@
+import React, { ComponentType, useEffect, useMemo, useState } from 'react';
+import {
+  rankWith,
+  ArrayControlProps,
+  findUISchema,
+  ControlElement,
+  composePaths,
+  uiTypeIs,
+  ControlProps,
+} from '@jsonforms/core';
+import {
+  withJsonFormsArrayControlProps,
+  JsonFormsDispatch,
+} from '@jsonforms/react';
+import { Typography } from '@openmsupply-client/common';
+
+import { JsonData } from '../../JsonForm';
+import { z } from 'zod';
+import { useZodOptionsValidation } from '../../hooks/useZodOptionsValidation';
+
+const Options = z
+  .object({
+    /** The field in the object which is used to extract the keyValue */
+    keyField: z.string(),
+    keyValue: z.string(),
+    detail: z.any().optional(),
+  })
+  .strict();
+type Options = z.infer<typeof Options>;
+
+interface UISchemaWithCustomProps extends ControlElement {
+  defaultNewItem?: JsonData;
+  itemLabel?: string;
+}
+
+interface KeyedItemArrayControlCustomProps
+  extends ArrayControlProps,
+    ControlProps {
+  uischema: UISchemaWithCustomProps;
+  data: JsonData[];
+  options: Options;
+}
+
+export const keyedItemArrayTester = rankWith(10, uiTypeIs('KeyedItemArray'));
+
+const KeyedItemArrayComponent: ComponentType<
+  KeyedItemArrayControlCustomProps
+> = (props: KeyedItemArrayControlCustomProps) => {
+  const {
+    uischema,
+    uischemas,
+    schema,
+    path,
+    enabled,
+    visible,
+    rootSchema,
+    renderers,
+    data,
+    handleChange,
+  } = props;
+
+  const [index, setIndex] = useState<number>(-1);
+  const { errors: zErrors, options } = useZodOptionsValidation(
+    Options,
+    uischema.options
+  );
+
+  useEffect(() => {
+    if (!options) return;
+    const arrayData = data ?? [];
+    const elementIndex = arrayData.findIndex(it => {
+      if (!it || typeof it !== 'object' || Array.isArray(it)) return;
```

Comment:
nit: can `it`s be renamed?

Reply:
Got this from Kotlin, they have an implicit `it` parameter in lambdas... Can call it `item`. Like `it` though, `item` or `element` are IMHO not much better but longer...